File: 2025/07 July/11 Trail of ones/main.py
# Link: https://www.geeksforgeeks.org/problems/trail-of-ones3242/1


class Solution:
    def countConsec(self, n: int) -> int:
        prev0, prev1 = 0, 0

        for i in range(n, 0, -1):
            curr0 = prev0 + prev1
            curr1 = prev0 + (1 << (n - i))
            prev0, prev1 = curr0, curr1

        return prev0


if __name__ == "__main__":
    sol = Solution()
    n = 2
    result = sol.countConsec(n)
    print(f"Number of ways to form a trail of ones for n = {n}: {result}")

    n = 3
    result = sol.countConsec(n)
    print(f"Number of ways to form a trail of ones for n = {n}: {result}")

    n = 5
    result = sol.countConsec(n)
    print(f"Number of ways to form a trail of ones for n = {n}: {result}")

File: 2025/07 July/11 Trail of ones/main.cpp
// Link: https://www.geeksforgeeks.org/problems/trail-of-ones3242/1

#include <iostream>
using namespace std;

class Solution
{
public:
    int countConsec(int n)
    {
        int prev0 = 0, prev1 = 0;
        for (int i = n; i >= 1; i--)
        {
            int curr0 = prev0 + prev1;
            int curr1 = prev0 + (1 << (n - i));
            prev0 = curr0;
            prev1 = curr1;
        }

        return prev0;
    }
};

int main()
{
    Solution sol;
    int n = 2;
    int result = sol.countConsec(n);
    cout << "Number of ways to form a trail of ones for n = " << n << " is: " << result << endl;

    n = 3;
    result = sol.countConsec(n);
    cout << "Number of ways to form a trail of ones for n = " << n << " is: " << result << endl;

    n = 5;
    result = sol.countConsec(n);
    cout << "Number of ways to form a trail of ones for n = " << n << " is: " << result << endl;

    return 0;
}